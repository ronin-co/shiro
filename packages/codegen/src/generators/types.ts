import { SyntaxKind, addSyntheticLeadingComment, factory } from 'typescript';

import { genericIdentifiers, identifiers } from '@/src/constants/identifiers';
import { convertToPascalCase } from '@/src/utils/slug';
import { mapRoninFieldToTypeNode, remapNestedFields } from '@/src/utils/types';

import type {
  InterfaceDeclaration,
  TypeAliasDeclaration,
  TypeParameterDeclaration,
} from 'typescript';

import type { Model, ModelField } from '@/src/types/model';

const DEFAULT_FIELD_SLUGS = [
  'id',
  'ronin.createdAt',
  'ronin.createdBy',
  'ronin.locked',
  'ronin.updatedAt',
  'ronin.updatedBy',
] satisfies Array<string>;

/**
 * Generate all required type definitions for a provided RONIN model.
 *
 * This will generate a shared schema interface that is then used to create type
 * aliases for both the singular and plural model types.
 *
 * The plural model type will be mapped to an array of the singular model type
 * and extend it with the plural model properties.
 *
 * @param models - All RONIN models of the addressed space.
 *
 * @returns - An array of type nodes to be added to the `index.d.ts` file.
 */
export const generateTypes = (
  models: Array<Model>,
): Array<InterfaceDeclaration | TypeAliasDeclaration> => {
  const nodes = new Array<InterfaceDeclaration | TypeAliasDeclaration>();

  for (const model of models) {
    const fields: Array<ModelField> = Object.entries(model.fields)
      .map(([slug, field]) => ({ ...field, slug }) as ModelField)
      .filter((field) => !DEFAULT_FIELD_SLUGS.includes(field.slug));

    const singularModelIdentifier = factory.createIdentifier(
      convertToPascalCase(model.slug),
    );
    const pluralSchemaIdentifier = factory.createIdentifier(
      convertToPascalCase(model.pluralSlug),
    );

    const hasLinkFields = fields.some(
      (field) =>
        field.type === 'link' && models.some((model) => model.slug === field.target),
    );

    const modelInterfaceTypeParameters = new Array<TypeParameterDeclaration>();
    const linkFieldKeys = fields
      .filter((field) => field.type === 'link')
      .map((field) => {
        const literal = factory.createStringLiteral(field.slug);
        return factory.createLiteralTypeNode(literal);
      });

    /**
     * ```ts
     * <TUsing extends Array<...> | 'all' = []>
     * ```
     */
    const usingGenericDec = factory.createTypeParameterDeclaration(
      undefined,
      genericIdentifiers.using,
      factory.createUnionTypeNode([
        factory.createTypeReferenceNode(identifiers.primitive.array, [
          factory.createUnionTypeNode(linkFieldKeys),
        ]),
        factory.createLiteralTypeNode(factory.createStringLiteral('all')),
      ]),
      factory.createTupleTypeNode([]),
    );

    if (hasLinkFields) modelInterfaceTypeParameters.push(usingGenericDec);

    /**
     * ```ts
     * SchemaSlugSchema<TUsing>
     * ```
     */
    const modelSchemaName = factory.createTypeReferenceNode(
      singularModelIdentifier,
      hasLinkFields ? [factory.createTypeReferenceNode(genericIdentifiers.using)] : [],
    );

    /**
     * ```ts
     * export type SchemaSlug<TUsing extends Array<...> | 'all' = []> = ResultRecord & {
     *  // ...
     * };
     * ```
     */
    const singularModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      singularModelIdentifier,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        factory.createExpressionWithTypeArguments(
          identifiers.syntax.resultRecord,
          undefined,
        ),
        factory.createTypeLiteralNode(
          remapNestedFields(fields)
            .sort(([a], [b]) => a.localeCompare(b))
            .map(([fieldSlug, field]) => {
              if (Array.isArray(field)) {
                const sortedFields = field.sort((a, b) => a.slug.localeCompare(b.slug));
                return factory.createPropertySignature(
                  undefined,
                  fieldSlug,
                  undefined,
                  factory.createTypeLiteralNode(
                    sortedFields.map((nestedField) =>
                      factory.createPropertySignature(
                        undefined,
                        nestedField.slug,
                        undefined,
                        factory.createUnionTypeNode(
                          mapRoninFieldToTypeNode(nestedField, models),
                        ),
                      ),
                    ),
                  ),
                );
              }

              return factory.createPropertySignature(
                undefined,
                fieldSlug,
                undefined,
                factory.createUnionTypeNode(mapRoninFieldToTypeNode(field, models)),
              );
            }),
        ),
      ]),
    );

    /**
     * ```ts
     * Array<SchemaSlug>;
     * ```
     */
    const pluralModelArrayTypeDec = factory.createTypeReferenceNode(
      identifiers.primitive.array,
      [modelSchemaName],
    );

    /**
     * ```ts
     * {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelPaginationPropsTypeDec = factory.createTypeLiteralNode([
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreBefore'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier('moreAfter'),
        factory.createToken(SyntaxKind.QuestionToken),
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
      ),
    ]);

    /**
     * ```ts
     * export type SchemaPluralSlug<TUsing extends Array<...> | 'all' = []> = Array<SchemaSlug> & {
     *  moreBefore?: string;
     *  moreAfter?: string;
     * };
     * ```
     */
    const pluralModelTypeDec = factory.createTypeAliasDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      pluralSchemaIdentifier,
      modelInterfaceTypeParameters,
      factory.createIntersectionTypeNode([
        pluralModelArrayTypeDec,
        pluralModelPaginationPropsTypeDec,
      ]),
    );

    // If the model does not have a summary / description
    // then we can continue to the next iteration & not add any comments.
    if (!model.summary) {
      nodes.push(singularModelTypeDec, pluralModelTypeDec);
      continue;
    }

    nodes.push(
      addSyntheticLeadingComment(
        singularModelTypeDec,
        SyntaxKind.MultiLineCommentTrivia,
        `*\n * ${model.summary}\n `,
        true,
      ),
      addSyntheticLeadingComment(
        pluralModelTypeDec,
        SyntaxKind.MultiLineCommentTrivia,
        `*\n * ${model.summary}\n `,
        true,
      ),
    );
  }

  return nodes;
};
