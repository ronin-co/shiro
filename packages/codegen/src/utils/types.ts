import { SyntaxKind, factory } from 'typescript';

import { genericIdentifiers, identifiers } from '@/src/constants/identifiers';
import { MODEL_TYPE_TO_SYNTAX_KIND_KEYWORD } from '@/src/constants/schema';
import { convertToPascalCase } from '@/src/utils/slug';

import type { ModelField } from '@ronin/compiler';
import type { TypeNode } from 'typescript';

import type { Model } from '@/src/types/model';

/**
 * Map a RONIN model field to a TypeScript type node.
 *
 * @param field - The RONIN model field to map.
 * @param models - The list of all RONIN models. Used to resolve link fields.
 *
 * @returns An array of TypeScript type nodes representing the field type.
 */
export const mapRoninFieldToTypeNode = (
  field: ModelField,
  models: Array<Model>,
): Array<TypeNode> => {
  const propertyUnionTypes = new Array<TypeNode>();

  switch (field.type) {
    case 'link': {
      // Check to make sure the target model exists. If it doesn't we
      // fall back to using `unknown` as the type.
      const targetModel = models.find((model) => model.slug === field.target);
      if (!targetModel) {
        propertyUnionTypes.push(factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword));
        break;
      }

      // If the field is marked as `many` then we need to wrap the
      // type in an array.
      const schemaTypeRef = factory.createTypeReferenceNode(
        convertToPascalCase(targetModel.slug),
      );
      const resolvedLinkFieldNode = factory.createTypeReferenceNode(
        identifiers.utils.resolveSchema,
        [
          field.kind === 'many'
            ? factory.createTypeReferenceNode(identifiers.primitive.array, [
                schemaTypeRef,
              ])
            : schemaTypeRef,

          factory.createTypeReferenceNode(genericIdentifiers.using),

          factory.createLiteralTypeNode(factory.createStringLiteral(field.slug)),
        ],
      );

      propertyUnionTypes.push(resolvedLinkFieldNode);
      break;
    }
    case 'blob':
    case 'boolean':
    case 'date':
    case 'json':
    case 'number':
    case 'string': {
      const primitive = MODEL_TYPE_TO_SYNTAX_KIND_KEYWORD[field.type];
      propertyUnionTypes.push(primitive);
      break;
    }
    default: {
      propertyUnionTypes.push(factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword));
      break;
    }
  }

  // We need to mark fields as nullable if they are:
  // - Not required
  // - Not a link field
  // - Not a many-to-many link field
  if (field.required === false && field.type === 'link' && field.kind !== 'many')
    propertyUnionTypes.push(factory.createLiteralTypeNode(factory.createNull()));

  return propertyUnionTypes;
};

/**
 * Remap nested fields to a more usable format.
 *
 * Currently, nested fields for a RONIN model are stored using a dot-notation
 * slug, such as `foo.bar`. However, these are resolved as a nested object.
 * As such, we need to remap the fields to a more usable format.
 *
 * @param fields - The list of fields to remap.
 *
 * @returns An array of entries where the key is the field slug and the value is
 * either the field itself or an array of (nested) fields.
 */
export const remapNestedFields = (
  fields: Array<ModelField>,
): Array<[string, ModelField | Array<ModelField>]> => {
  const remappedFields = new Map<string, ModelField | Array<ModelField>>();

  for (const field of fields) {
    if (!field.slug.includes('.')) {
      remappedFields.set(field.slug, field);
      continue;
    }

    const [parentSlug, childSlug] = field.slug.split('.');
    const nestedField = Object.assign({}, field, {
      slug: childSlug,
    });

    const parentField = remappedFields.get(parentSlug);
    if (!parentField) {
      remappedFields.set(parentSlug, [nestedField]);
      continue;
    }

    if (!Array.isArray(parentField)) continue;

    parentField.push(nestedField);
  }

  return Array.from(remappedFields.entries());
};
