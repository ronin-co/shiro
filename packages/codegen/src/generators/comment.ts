import { READABLE_DML_QUERY_TYPES } from '@/src/constants/schema';

import type { Model } from '@/src/types/model';
import type { DML_QUERY_TYPES } from '@ronin/compiler';

interface GenerateQueryTypeCommentResult {
  singular: string;
  plural: string;
}

/**
 * Generate a text comment for a provided model & a given query type.
 *
 * @param modelName - The name of the model to generate a comment for.
 * @param queryType - The query type to generate a comment for.
 *
 * @returns An object containing both the singular and plural comment strings.
 */
export const generateQueryTypeComment = (
  model: Model,
  queryType: (typeof DML_QUERY_TYPES)[number],
): GenerateQueryTypeCommentResult => ({
  singular: ` ${READABLE_DML_QUERY_TYPES[queryType]} a single ${model.name ?? model.slug} record `,
  plural: ` ${READABLE_DML_QUERY_TYPES[queryType]} multiple ${model.name ?? model.slug} records `,
});
