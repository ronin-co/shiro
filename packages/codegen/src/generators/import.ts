import { factory } from 'typescript';

import type { Expression, Identifier, ImportDeclaration } from 'typescript';

interface CreateImportDeclarationOptions {
  /**
   * A list of all the identifiers that should be imported.
   */
  identifiers: Array<{
    alias?: Identifier;
    /**
     * The name of the identifier to import.
     */
    name: Identifier;
    /**
     * Whether the identifier should be marked as a type import.
     */
    type?: boolean;
  }>;
  /**
   * The name of the module or package or path to import from.
   */
  module: Expression;
  /**
   * Whether the import should be marked as a type import.
   */
  type?: boolean;
}

/**
 * Generates an `import {} from 'foobar';` declaration using a provided list of
 * identifiers.
 *
 * @param options - The options to use when generating the import declaration.
 *
 * @returns The generated import declaration.
 *
 * @example
 * ```ts
 * import { factory } from 'typescript';
 *
 * const declaration = createImportDeclaration({
 *  identifiers: [{ name: factory.createIdentifier('RONIN') }],
 *  module: factory.createIdentifier('ronin'),
 *  type: true,
 * });
 * // import type { RONIN } from 'ronin';
 * ```
 */
export const createImportDeclaration = (
  options: CreateImportDeclarationOptions,
): ImportDeclaration => {
  const namedBindings = factory.createNamedImports(
    options.identifiers.map((identifier) =>
      factory.createImportSpecifier(
        identifier.type ?? false,
        identifier.alias ? identifier.name : undefined,
        identifier.alias ? identifier.alias : identifier.name,
      ),
    ),
  );

  const importClause = factory.createImportClause(
    options.type ?? false,
    undefined,
    namedBindings,
  );

  return factory.createImportDeclaration(undefined, importClause, options.module);
};
