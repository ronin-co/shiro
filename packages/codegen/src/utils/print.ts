import {
  ListFormat,
  NewLineKind,
  ScriptKind,
  ScriptTarget,
  createPrinter,
  createSourceFile,
  factory,
} from 'typescript';

import { DEFAULT_FILE_NAME } from '@/src/constants/file';

import type { Node } from 'typescript';

interface PrintNodesOptions {
  /**
   * The name of the source file.
   *
   * @default 'index.d.ts'
   */
  fileName?: string;
  /**
   * Whether to remove comments from the output.
   *
   * @default false
   */
  removeComments?: boolean;
}

/**
 * Prints a list of nodes to a string.
 *
 * This is primarily used to actually output code pragmatically
 * generated to a string.
 *
 * @param nodes - The nodes to print.
 *
 * @returns The string representation of the nodes.
 */
export const printNodes = (nodes: Array<Node>, options?: PrintNodesOptions): string => {
  const { fileName = DEFAULT_FILE_NAME, removeComments = false } = options ?? {};

  // This is a mock file that is entirely in memory & blank. It is just
  // required for the printer to work. The printer will add the nodes to
  // this file & then return the string representation of the file.
  const sourceFile = createSourceFile(
    fileName,
    '',
    ScriptTarget.ESNext,
    true,
    ScriptKind.TS,
  );

  const printer = createPrinter({
    newLine: NewLineKind.LineFeed,
    removeComments,
  });

  const nodesArray = factory.createNodeArray(nodes);

  return printer.printList(ListFormat.MultiLine, nodesArray, sourceFile);
};
