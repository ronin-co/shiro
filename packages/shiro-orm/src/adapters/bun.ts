import { Database } from 'bun:sqlite';

import type { Model } from 'shiro-compiler';

interface ShiroOptions<T extends Array<Model>> {
  database?: Database;
  models: T;
}

/**
 * @todo Add documentation
 */
export const shiro = <T extends Array<Model>>(options: ShiroOptions<T>) => {
  const database = options.database ?? new Database(':memory:');
};
