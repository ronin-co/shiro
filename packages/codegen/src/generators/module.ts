import { DDL_QUERY_TYPES, DML_QUERY_TYPES } from 'shiro-compiler';
import { NodeFlags, SyntaxKind, addSyntheticLeadingComment, factory } from 'typescript';

import { genericIdentifiers, identifiers } from '@/src/constants/identifiers';
import { generateQueryTypeComment } from '@/src/generators/comment';
import { convertToPascalCase } from '@/src/utils/slug';

import type {
  InterfaceDeclaration,
  ModuleDeclaration,
  ParameterDeclaration,
  PropertySignature,
  Statement,
  TypeAliasDeclaration,
  TypeElement,
} from 'typescript';

import type { Model } from '@/src/types/model';

/**
 * Generate a module augmentation for the `ronin` module to override the
 * standard filter interfaces with ones that are correctly typed specific to
 * this space.
 *
 * @param models - An array of RONIN models to generate type definitions for.
 * @param schemas - An array of type declarations for the models.
 *
 * @returns A module augmentation declaration to be added to `index.d.ts`.
 */
export const generateModule = (
  models: Array<Model>,
  schemas: Array<InterfaceDeclaration | TypeAliasDeclaration>,
): ModuleDeclaration => {
  const moduleBodyStatements = new Array<Statement>();

  for (const schemaTypeDec of schemas) {
    moduleBodyStatements.push(schemaTypeDec);
  }

  const mappedQueryTypeVariableDeclarations = DML_QUERY_TYPES.map((queryType) => {
    const declarationProperties = new Array<TypeElement>();

    for (const model of models) {
      const comment = generateQueryTypeComment(model, queryType);
      const singularModelIdentifier = factory.createTypeReferenceNode(
        convertToPascalCase(model.slug),
      );
      const pluralSchemaIdentifier = factory.createTypeReferenceNode(
        convertToPascalCase(model.pluralSlug),
      );

      /**
       * ```ts
       * GetQuery[keyof GetQuery]
       * ```
       */
      const queryTypeValue = factory.createIndexedAccessTypeNode(
        factory.createTypeReferenceNode(
          identifiers.compiler.dmlQueryType[queryType],
          undefined,
        ),
        factory.createTypeOperatorNode(
          SyntaxKind.KeyOfKeyword,
          factory.createTypeReferenceNode(identifiers.compiler.dmlQueryType[queryType]),
        ),
      );

      /**
       * ```ts
       * account: DeepCallable<GetQuery[keyof GetQuery], Account | null>;
       * ```
       */
      const singularProperty = factory.createPropertySignature(
        undefined,
        model.slug,
        undefined,
        factory.createTypeReferenceNode(identifiers.syntax.deepCallable, [
          queryTypeValue,
          factory.createUnionTypeNode(
            queryType === 'count'
              ? [factory.createKeywordTypeNode(SyntaxKind.NumberKeyword)]
              : [
                  singularModelIdentifier,
                  factory.createLiteralTypeNode(factory.createNull()),
                ],
          ),
        ]),
      );

      // There is no value in supporting `count` queries for singular
      // records, so we skip adding the comment for those.
      if (queryType !== 'count')
        declarationProperties.push(
          addSyntheticLeadingComment(
            singularProperty,
            SyntaxKind.MultiLineCommentTrivia,
            comment.singular,
            true,
          ),
        );

      // TODO(@nurodev): Remove once RONIN officially supports
      // creating multiple records at once.
      if (queryType === 'add') continue;

      /**
       * ```ts
       * accounts: DeepCallable<GetQuery[keyof GetQuery], Array<Account>>;
       * ```
       */
      const pluralProperty = factory.createPropertySignature(
        undefined,
        model.pluralSlug,
        undefined,
        factory.createTypeReferenceNode(identifiers.syntax.deepCallable, [
          queryTypeValue,
          queryType === 'count'
            ? factory.createKeywordTypeNode(SyntaxKind.NumberKeyword)
            : pluralSchemaIdentifier,
        ]),
      );
      declarationProperties.push(
        addSyntheticLeadingComment(
          pluralProperty,
          SyntaxKind.MultiLineCommentTrivia,
          comment.plural,
          true,
        ),
      );
    }

    return {
      properties: declarationProperties,
      queryType,
    };
  });

  /**
   * ```ts
   * declare const add: { ... };
   * declare const count: { ... };
   * declare const get: { ... };
   * declare const remove: { ... };
   * declare const set: { ... };
   * ```
   */
  for (const { properties, queryType } of mappedQueryTypeVariableDeclarations) {
    const queryDeclaration = factory.createVariableStatement(
      [factory.createModifier(SyntaxKind.DeclareKeyword)],
      factory.createVariableDeclarationList(
        [
          factory.createVariableDeclaration(
            queryType,
            undefined,
            factory.createTypeLiteralNode(properties),
          ),
        ],
        NodeFlags.Const,
      ),
    );

    moduleBodyStatements.push(queryDeclaration);
  }

  /**
   * ```ts
   * T extends [Promise, ...Array<Promise>] | Array<Promise>
   * ```
   */
  const batchQueryTypeArguments = factory.createTypeParameterDeclaration(
    undefined,
    genericIdentifiers.queries,
    factory.createUnionTypeNode([
      factory.createTupleTypeNode([
        factory.createTypeReferenceNode(identifiers.primitive.promise),
        factory.createRestTypeNode(
          factory.createTypeReferenceNode(identifiers.primitive.array, [
            factory.createTypeReferenceNode(identifiers.primitive.promise),
          ]),
        ),
      ]),

      factory.createTypeReferenceNode(identifiers.primitive.array, [
        factory.createTypeReferenceNode(identifiers.primitive.promise),
      ]),
    ]),
  );

  const batchQueryParametersDeclaration = new Array<ParameterDeclaration>();

  /**
   * ```ts
   * operations: () => T
   * ```
   */
  batchQueryParametersDeclaration.push(
    factory.createParameterDeclaration(
      undefined,
      undefined,
      'operations',
      undefined,
      factory.createFunctionTypeNode(
        undefined,
        [],
        factory.createTypeReferenceNode(genericIdentifiers.queries),
      ),
    ),
  );

  /**
   * ```ts
   * queryOptions?: Record<string, unknown>
   * ```
   */
  batchQueryParametersDeclaration.push(
    factory.createParameterDeclaration(
      undefined,
      undefined,
      'queryOptions',
      factory.createToken(SyntaxKind.QuestionToken),
      factory.createTypeReferenceNode(identifiers.primitive.record, [
        factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
        factory.createKeywordTypeNode(SyntaxKind.UnknownKeyword),
      ]),
    ),
  );

  /**
   * ```ts
   * declare const batch: <...>(...) => Promise<PromiseTuple<T>>;
   * ```
   */
  const batchQueryDeclaration = factory.createVariableStatement(
    [factory.createModifier(SyntaxKind.DeclareKeyword)],
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          'batch',
          undefined,
          factory.createFunctionTypeNode(
            [batchQueryTypeArguments],
            batchQueryParametersDeclaration,
            factory.createTypeReferenceNode(identifiers.primitive.promise, [
              factory.createTypeReferenceNode(identifiers.ronin.promiseTuple, [
                factory.createTypeReferenceNode(genericIdentifiers.queries),
              ]),
            ]),
          ),
        ),
      ],
      NodeFlags.Const,
    ),
  );
  moduleBodyStatements.push(batchQueryDeclaration);

  /**
   * ```ts
   * models: DeepCallable<ListQuery[keyof ListQuery], Array<Model>>;
   * ```
   */
  const listModelsQueryPropertyDeclaration = addSyntheticLeadingComment(
    factory.createPropertySignature(
      undefined,
      'models',
      undefined,
      factory.createTypeReferenceNode(identifiers.syntax.deepCallable, [
        factory.createIndexedAccessTypeNode(
          factory.createTypeReferenceNode(
            identifiers.compiler.ddlQueryType.list,
            undefined,
          ),
          factory.createTypeOperatorNode(
            SyntaxKind.KeyOfKeyword,
            factory.createTypeReferenceNode(identifiers.compiler.ddlQueryType.list),
          ),
        ),
        factory.createTypeReferenceNode(identifiers.primitive.array, [
          factory.createTypeReferenceNode(identifiers.compiler.model),
        ]),
      ]),
    ),
    SyntaxKind.MultiLineCommentTrivia,
    ' List all model definitions ',
    true,
  );

  /**
   * ```ts
   * declare const list: { ... };
   * ```
   */
  const listModelsQueryDeclaration = factory.createVariableStatement(
    [factory.createModifier(SyntaxKind.DeclareKeyword)],
    factory.createVariableDeclarationList(
      [
        factory.createVariableDeclaration(
          'list',
          undefined,
          factory.createTypeLiteralNode([listModelsQueryPropertyDeclaration]),
        ),
      ],
      NodeFlags.Const,
    ),
  );

  moduleBodyStatements.push(listModelsQueryDeclaration);

  // Note: `csf` prefix stands for `createSyntaxFactory`.

  /**
   * ```ts
   * (options: QueryHandlerOptions | (() => QueryHandlerOptions))
   * ```
   */
  const csfParameterTypeDec = factory.createParameterDeclaration(
    undefined,
    undefined,
    'options',
    undefined,
    factory.createUnionTypeNode([
      factory.createTypeReferenceNode(identifiers.ronin.queryHandlerOptions),
      factory.createFunctionTypeNode(
        undefined,
        [],
        factory.createTypeReferenceNode(identifiers.ronin.queryHandlerOptions),
      ),
    ]),
  );

  const csfReturnTypePropertySignatures = new Array<PropertySignature>();

  /**
   * ```ts
   * (...) => {
   *  add: typeof add,
   *  count: typeof count,
   *  get: typeof get,
   *  remove: typeof remove,
   *  set: typeof set,
   *  list: typeof list,
   * }
   * ```
   */
  for (const queryType of [...DML_QUERY_TYPES, 'list']) {
    csfReturnTypePropertySignatures.push(
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier(queryType),
        undefined,
        factory.createTypeQueryNode(factory.createIdentifier(queryType)),
      ),
    );
  }

  /**
   * ```ts
   * (...) => {
   *  create: typeof import('ronin').create,
   *  alter: typeof import('ronin').alter,
   *  drop: typeof import('ronin').drop,
   *  batch: typeof import('ronin').batch,
   *  sql: typeof import('ronin').sql,
   *  sqlBatch: typeof import('ronin').sqlBatch,
   * }
   * ```
   */
  for (const queryType of [
    ...DDL_QUERY_TYPES.filter((v) => v !== 'list'),
    'batch',
    'sql',
    'sqlBatch',
  ]) {
    csfReturnTypePropertySignatures.push(
      factory.createPropertySignature(
        undefined,
        factory.createIdentifier(queryType),
        undefined,
        factory.createTypeQueryNode(
          factory.createQualifiedName(
            // Currently this is the only viable option I have found to implement a
            // format of `import('ronin').xyz` node in the TSC API.
            // But with this the TSC API marks these properties as not compatible,
            // but pragmatically they work fine.
            // @ts-expect-error
            factory.createImportTypeNode(
              factory.createTypeReferenceNode(identifiers.ronin.module.root),
            ),
            factory.createIdentifier(queryType),
          ),
        ),
      ),
    );
  }

  const csfReturnTypeDec = factory.createTypeLiteralNode(csfReturnTypePropertySignatures);

  moduleBodyStatements.push(
    /**
     * ```ts
     * declare const createSyntaxFactory: (
     *  options: QueryHandlerOptions | (() => QueryHandlerOptions)
     * ) => { ... }
     * ```
     */
    factory.createVariableStatement(
      [factory.createModifier(SyntaxKind.DeclareKeyword)],
      factory.createVariableDeclarationList(
        [
          factory.createVariableDeclaration(
            identifiers.ronin.createSyntaxFactory,
            undefined,
            factory.createFunctionTypeNode(
              undefined,
              [csfParameterTypeDec],
              csfReturnTypeDec,
            ),
          ),
        ],
        NodeFlags.Const,
      ),
    ),

    /**
     * ```ts
     * export default function (
     *  options: QueryHandlerOptions | (() => QueryHandlerOptions)
     * ) => { ... }
     * ```
     */
    factory.createFunctionDeclaration(
      [
        factory.createModifier(SyntaxKind.ExportKeyword),
        factory.createModifier(SyntaxKind.DefaultKeyword),
      ],
      undefined,
      undefined,
      undefined,
      [csfParameterTypeDec],
      csfReturnTypeDec,
      undefined,
    ),
  );

  return factory.createModuleDeclaration(
    [factory.createModifier(SyntaxKind.DeclareKeyword)],
    identifiers.ronin.module.root,
    factory.createModuleBlock(moduleBodyStatements),
  );
};
