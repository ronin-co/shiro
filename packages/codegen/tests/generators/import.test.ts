import { describe, expect, test } from 'bun:test';
import { factory } from 'typescript';

import { createImportDeclaration } from '@/src/generators/import';
import { printNodes } from '@/src/utils/print';

// Note: We need to `JSON.stringify` the module name to ensure that the
// module name is wrapped in quotes for the import statement.
const roninModuleIdentifier = factory.createIdentifier(JSON.stringify('ronin'));
const roninNamespaceIdentifier = factory.createIdentifier('RONIN');

describe('import', () => {
  test('a simple import statement', () => {
    const declaration = createImportDeclaration({
      identifiers: [{ name: roninNamespaceIdentifier }],
      module: roninModuleIdentifier,
    });

    const output = printNodes([declaration]);

    expect(output).toMatchSnapshot();
  });

  test('a type import statement', () => {
    const declaration = createImportDeclaration({
      identifiers: [{ name: roninNamespaceIdentifier }],
      module: roninModuleIdentifier,
      type: true,
    });

    const output = printNodes([declaration]);

    expect(output).toMatchSnapshot();
  });

  test('an import statement with individual type imports', () => {
    const declaration = createImportDeclaration({
      identifiers: [{ name: roninNamespaceIdentifier, type: true }],
      module: roninModuleIdentifier,
    });

    const output = printNodes([declaration]);

    expect(output).toMatchSnapshot();
  });

  test('an import with multiple identifiers', () => {
    const declaration = createImportDeclaration({
      identifiers: [
        { name: factory.createIdentifier('Foo') },
        { name: factory.createIdentifier('Bar') },
      ],
      module: roninModuleIdentifier,
    });

    const output = printNodes([declaration]);

    expect(output).toMatchSnapshot();
  });

  test('an import with mixed typed identifiers', () => {
    const declaration = createImportDeclaration({
      identifiers: [
        { name: factory.createIdentifier('Foo') },
        { name: factory.createIdentifier('Bar'), type: true },
      ],
      module: roninModuleIdentifier,
    });

    const output = printNodes([declaration]);

    expect(output).toMatchSnapshot();
  });
});
