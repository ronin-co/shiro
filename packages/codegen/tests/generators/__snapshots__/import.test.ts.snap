// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`import a simple import statement 1`] = `
"import { RONIN } from "shiro-orm";
"
`;

exports[`import a type import statement 1`] = `
"import type { RONIN } from "shiro-orm";
"
`;

exports[`import an import statement with individual type imports 1`] = `
"import { type RONIN } from "shiro-orm";
"
`;

exports[`import an import with multiple identifiers 1`] = `
"import { Foo, Bar } from "shiro-orm";
"
`;

exports[`import an import with mixed typed identifiers 1`] = `
"import { Foo, type Bar } from "shiro-orm";
"
`;
